Add PREDICTION module

vodogrey2012 opened the pull request:
Good night.
I have tried to solve this issue
https://github.com/MIPT-ILab/mipt-mips/issues/724

I created a separate module PREDICTOR, before it was part of the MEM module.
For this, I created another port in EXECUTE.

Checked, IPS has not changed.

vodogrey2012 committed:
Add PREDICTION module

Create new separate PREDICTION module, before it was a part of MEM module.
Also, changed CMake.

vodogrey2012 committed:
Add PREDICTION module

Create new separate PREDICTION module, before it was a part of MEM module.
Also, changed CMake.

vodogrey2012 commented:
Also, i changed CMakeList.txt file.

vodogrey2012 committed:
Change solved style problems

vodogrey2012 committed:
Remove unused variable

vodogrey2012 committed:
Solve style problem

pavelkryukov commented on simulator/CMakeLists.txt:
```diff
@@ -35,6 +35,7 @@ set(CPPS infra/macro_test.cpp
     modules/decode/decode.cpp
     modules/execute/execute.cpp
     modules/mem/mem.cpp
+    modules/prediction/prediction.cpp
```
The name is misleading — it does not predict anything. I suggest to rename it to `branch`, including classes and objects.

pavelkryukov commented on simulator/modules/execute/execute.cpp:
```diff
@@ -123,7 +124,10 @@ void Execute<ISA>::clock( Cycle cycle)
         wp_bypass->write( instr.get_dst_v(), cycle);
 
         if ( instr.is_mem_stage_required())
+	{
```
Please, do not use tabs for indentation here and everywhere.

pavelkryukov replied:
Please double-check for tabs and set up your editor: https://github.com/MIPT-ILab/mipt-mips/wiki/Code-style-guidelines#indentation

pavelkryukov commented on simulator/modules/execute/execute.cpp:
```diff
@@ -123,7 +124,10 @@ void Execute<ISA>::clock( Cycle cycle)
 
         if ( instr.is_mem_stage_required())
+	{
             wp_mem_datapath->write( std::move( instr), cycle);
```
By the way, you may steer here — send jumps to `branch` port, and send loads and stores to `memory` port.

vodogrey2012 replied:
Excuse me, this means that I need to add a comment?) If not, can you describe in more detail? Do I need to change data, which is sent to MEM and BRANCH?

pavelkryukov replied:
I think you can do this:
```c++
if ( instr.is_jump())
wp_prediction_datapath->write( std::move( instr) ,cycle);
else if ( instr.is_mem_stage_required())
    wp_mem_datapath->write( std::move( instr), cycle);
else
    wp_writeback_datapath->write( std::move( instr) , cycle)
```

vodogrey2012 replied:
I got this idea, thank you

vodogrey2012 replied:
I think, that it should be like this:
```ruby
if ( instr.is_jump())
    wp_branch_datapath->write( std::move( instr) ,cycle);
if ( instr.is_mem_stage_required())
    wp_mem_datapath->write( std::move( instr), cycle);
else
    wp_writeback_datapath->write( std::move( instr) , cycle)
```
because if branch was predicted correct, mem should be writing

pavelkryukov replied:
I think it contains a bug. At least, you cannot do `std::move( instr)` here anе use `instr` — that's not allowed by C++.
(If you wish, you may remove all `std::move`, I'll place them later myself, it's quite complicated language feature).

>  if branch was predicted correct, mem should be writing

Not sure I understand you. What should Mem write?

vodogrey2012 replied:
Good night.
> Not sure I understand you. What should Mem write?

I was wrong, jump doesn't load any memory. 
I have found information about `std::move( instr)` and it is really bug and must be solved.
I try to insert this variant
``` c++
if ( instr.is_jump())
wp_prediction_datapath->write( std::move( instr) ,cycle);
else if ( instr.is_mem_stage_required())
    wp_mem_datapath->write( std::move( instr), cycle);
else
    wp_writeback_datapath->write( std::move( instr) , cycle)
```
and unfortunately it always cause this
``` shell
Checker mismatch:
	Checker output: {3}	0x4000fc: bne $t0, $zero, 46
PerfSim output: {4}	0x4001b8: lui $s0, 0x41	 [ $s0 = 0x410000 ]
```
I have found, that problem is in writeback.cpp module:
``` c++
    /* perform checks */
    checker.check( instr);

    /* update simulator cycles info */
    ++executed_instrs;
    last_writeback_cycle = cycle;
    if ( executed_instrs >= instrs_to_run || instr.is_halt())
        wp_halt->write( true, cycle);
```
This functions must be performed for jump (at least it doesn't work without).
Moreover, if I remove rest functions (writers to ports), it decrease IPC.
So, I think that both modules `MEM` and `WRITEBACK` or `BRANCH` and `WRITEBACK` must be performed.

I see one easy way of solving this problem: I need to add `BRANCH_2_WRITEBACK`  datapath port for they communication. So I can keep good variant of steering and execution of both mobules.
Also, I can make `EXECUTE_2_` port with 3 readers: `MEM` , `BRANCH` and `WRITEBACK`. I think this variant is more logic.
I will try to push last variant in hour.

But if it is still is not, what I need to do, I will be pleased to talk with you at closest seminar, if it is possible.
 
Thank you

pavelkryukov replied:
> Also, I can make EXECUTE_2_ port with 3 readers: MEM , BRANCH and WRITEBACK. I think this variant is more logic.

Looks right — I thought you done that already.

> But if it is still is not, what I need to do, I will be pleased to talk with you at closest seminar, if it is possible.

I am on business trip in Haifa, so I will not be able to deliver the last seminar, Kirill will do it.
You may talk to Oleg about branch prediction issues.

pavelkryukov commented on simulator/modules/prediction/prediction.cpp:
```diff
@@ -0,0 +1,83 @@
+/**
+ * rediction.cpp - Simulation of Branch Mispredict Detection stage
```
Please add your name and e-mail here

pavelkryukov commented on simulator/modules/prediction/prediction.cpp:
```diff
@@ -0,0 +1,83 @@
+template <typename ISA>
+Prediction<ISA>::Prediction( bool log) : Log( log)
+{
+    wp_flush_all = make_write_port<bool>("MEMORY_2_ALL_FLUSH", PORT_BW, FLUSHED_STAGES_NUM);
```
`MEMORY_2_ALL` is an obsolete name now — please use BRANCH_2_ALL, same for all others

pavelkryukov reviewed (changes requested):
Great work.
Please see my comments — they are mostly about code form and naming.

vodogrey2012 committed:
Rename .h and .cpp files from PREDICTION to BRANCH

vodogrey2012 committed:
Rename Prediction class to Branch class

vodogrey2012 committed:
Substitute TABs with spaces

vodogrey2012 committed:
Add author information

vodogrey2012 committed:
Rename prediction port to branch port

vodogrey2012 committed:
Replace TABs with spaces

vodogrey2012 committed:
Modified steering of branch and mem

Now, instr is sent to the brunch module only, if it is jump

vodogrey2012 committed:
Change oder of checking ports

vodogrey2012 commented:
Sorry, It has problems in -d mode.
I will solve it in the evening.

vodogrey2012 committed:
Change execute, mem and brunch ports

By the way, that writeback module must perform after both of brunch and mem modules, I substitute wp_mem_datapath in execute module with wp_mem_and_branch_datapath, which has two readers.
So, is_brunch() is checked in brunch module now.

vodogrey2012 committed:
Solve problem with log
